{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Retrieving data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Introduction\n",
    "\n",
    "- Repertoire sequence datasets are large in size.\n",
    "- Generated by a variety of platforms, e.g.\n",
    "  - SFF (Roche 454);\n",
    "  - Illumina (Native/SRF);\n",
    "  - PacBio."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Sequence Read Archive (SRA)\n",
    "\n",
    "- The Sequence Read Archive (SRA) stores raw sequence data from next-generation sequencing technologies.\n",
    "- The SRA is part of the international partnership of archives (INSDC) at the NCBI, the European Bioinformatics Institute and the DNA Database of Japan.\n",
    "  - Data submitted to any of the three organizations are shared among them.\n",
    "- These database differ:\n",
    "  - In the way that metadata are stored.\n",
    "  - In how data are retrieved.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## BioProject\n",
    "\n",
    "- Holds metadata on why the analysis was performed\n",
    "  - Project title and abstract\n",
    "  - Aims and objectives\n",
    "  - Organism(s) sequenced\n",
    "  - Optional: Funding sources, publications, etc.\n",
    "- This links to an *SRA Study* that provides technical details."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## BioSample\n",
    "\n",
    "- Holds metadata on what was sequenced.\n",
    "  - Descriptive sample information\n",
    "  - Often in tabular format\n",
    "  - Examples: Organism(s), age(s), gender(s), location data, cell line(s), etc.\n",
    "- Links to *SRA Sample* that provides technical details."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## SRA Experiment\n",
    "\n",
    "- A description of a sample-specific sequencing library\n",
    "- How the sequencing was performed\n",
    "  - Sequencing methods\n",
    "  - Kits used\n",
    "  - Instrument model(s)\n",
    "- Multiple Experiments can 'point' to a single Sample, but not vice-versa\n",
    "- An Experiment is required for\n",
    "  - Different sequencing or preparation methods\n",
    "  - Biological or technical replicates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## SRA Run\n",
    "\n",
    "- Runs link data files to an Experiment\n",
    "- This represents the raw data that we will need to access\n",
    "- Stored in a proprietary format that needs to be converted e.g. to fastq before analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Accessing SRA using the web\n",
    "\n",
    "- Located at [https://www.ncbi.nlm.nih.gov/sra/](https://www.ncbi.nlm.nih.gov/sra/)\n",
    "- Access any data type stored in the SRA independently of any other data type.\n",
    "- Access related data from other NCBI resources that are integrated with SRA.\n",
    "- Retrieve data based on ancillary information and/or sequence comparisons.\n",
    "- Review the descriptions of studies and experiments (metadata) independently of experimental data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Searching NCBI\n",
    "\n",
    "- Can start at multiple points\n",
    "  - Publication in `pubmed`\n",
    "  - Project in `bioproject`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Entrez Direct\n",
    "\n",
    "- While using a web browser is convenient for searching for data, it lacks reproducibility, and is hard to share with others\n",
    "- [Entrez Direct](https://www.ncbi.nlm.nih.gov/books/NBK179288/) (EDirect) is a method for accessing the NCBI's suite of interconnected databases from the command line, made up of a series of Perl scripts.\n",
    " - `esearch` performs a new Entrez search using terms in indexed fields.\n",
    " - `elink` looks up neighbors (within a database) or links (between databases).\n",
    " - `efilter` filters or restricts the results of a previous query.\n",
    " - `efetch` downloads records or reports in a designated format."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Other EDirect utilities\n",
    "\n",
    "- `einfo` obtains information on indexed fields in an Entrez database.\n",
    "- `epost` uploads unique identifiers (UIDs) or sequence accession numbers.\n",
    "- `nquire` sends a URL request to a web page or CGI service.\n",
    "- `xtract` converts XML into a table of data values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<ENTREZ_DIRECT>\n",
      "  <Db>bioproject</Db>\n",
      "  <WebEnv>NCID_1_77295837_130.14.18.34_9001_1476509839_1876082318_0MetA0_S_MegaStore_F_1</WebEnv>\n",
      "  <QueryKey>1</QueryKey>\n",
      "  <Count>239</Count>\n",
      "  <Step>1</Step>\n",
      "</ENTREZ_DIRECT>\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "esearch -db bioproject -query \"immunoglobulin\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## SRA Toolkit\n",
    "\n",
    "- The [SRA Toolkit](https://www.ncbi.nlm.nih.gov/books/NBK158900/) allows programmatic access to data in the SRA and conversion of the SRA format into other formats\n",
    "- Comprises of several 'data dump' programs\n",
    "  - `fastq-dump`: converts to fastq and fasta formats\n",
    "- When combined with EDirect, this allows one to download data in fastq format, via a list of SRR accessions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Example\n",
    "\n",
    "- Finding the data from [Jiang et al. 2013](https://www.ncbi.nlm.nih.gov/pubmed/23390249)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Searching Pubmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<ENTREZ_DIRECT>\n",
      "  <Db>pubmed</Db>\n",
      "  <WebEnv>NCID_1_904445_130.14.22.215_9001_1477917337_1277602431_0MetA0_S_MegaStore_F_1</WebEnv>\n",
      "  <QueryKey>1</QueryKey>\n",
      "  <Count>1</Count>\n",
      "  <Step>1</Step>\n",
      "</ENTREZ_DIRECT>\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "esearch -db pubmed -query \"Jiang Quake lineage influenza\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Getting the abstract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1. Sci Transl Med. 2013 Feb 6;5(171):171ra19. doi: 10.1126/scitranslmed.3004794.\n",
      "\n",
      "Lineage structure of the human antibody repertoire in response to influenza\n",
      "vaccination.\n",
      "\n",
      "Jiang N(1), He J, Weinstein JA, Penland L, Sasaki S, He XS, Dekker CL, Zheng NY, \n",
      "Huang M, Sullivan M, Wilson PC, Greenberg HB, Davis MM, Fisher DS, Quake SR.\n",
      "\n",
      "Author information: \n",
      "(1)Department of Bioengineering, Stanford University School of Medicine,\n",
      "Stanford, CA 94305, USA.\n",
      "\n",
      "Erratum in\n",
      "    Sci Transl Med. 2013 Jul 10;5(193):193er8.\n",
      "\n",
      "The human antibody repertoire is one of the most important defenses against\n",
      "infectious disease, and the development of vaccines has enabled the conferral of \n",
      "targeted protection to specific pathogens. However, there are many challenges to \n",
      "measuring and analyzing the immunoglobulin sequence repertoire, including that\n",
      "each B cell's genome encodes a distinct antibody sequence, that the antibody\n",
      "repertoire changes over time, and the high similarity between antibody sequences.\n",
      "We have addressed these challenges by using high-throughput long read sequencing \n",
      "to perform immunogenomic characterization of expressed human antibody repertoires\n",
      "in the context of influenza vaccination. Informatic analysis of 5 million\n",
      "antibody heavy chain sequences from healthy individuals allowed us to perform\n",
      "global characterizations of isotype distributions, determine the lineage\n",
      "structure of the repertoire, and measure age- and antigen-related mutational\n",
      "activity. Our analysis of the clonal structure and mutational distribution of\n",
      "individuals' repertoires shows that elderly subjects have a decreased number of\n",
      "lineages but an increased prevaccination mutation load in their repertoire and\n",
      "that some of these subjects have an oligoclonal character to their repertoire in \n",
      "which the diversity of the lineages is greatly reduced relative to younger\n",
      "subjects. We have thus shown that global analysis of the immune system's clonal\n",
      "structure provides direct insight into the effects of vaccination and provides a \n",
      "detailed molecular portrait of age-related effects.\n",
      "\n",
      "DOI: 10.1126/scitranslmed.3004794 \n",
      "PMCID: PMC3699344\n",
      "PMID: 23390249  [PubMed - indexed for MEDLINE]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "esearch -db pubmed -query \"Jiang Quake lineage influenza\" | efetch -format abstract"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Searching SRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<ENTREZ_DIRECT>\n",
      "  <Db>sra</Db>\n",
      "  <WebEnv>NCID_1_1433884_130.14.22.215_9001_1477919254_952558587_0MetA0_S_MegaStore_F_1</WebEnv>\n",
      "  <QueryKey>1</QueryKey>\n",
      "  <Count>1</Count>\n",
      "  <Step>1</Step>\n",
      "</ENTREZ_DIRECT>\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "esearch -db sra -query SRA058972"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run,ReleaseDate,LoadDate,spots,bases,spots_with_mates,avgLength,size_MB,AssemblyName,download_path,Experiment,LibraryName,LibraryStrategy,LibrarySelection,LibrarySource,LibraryLayout,InsertSize,InsertDev,Platform,Model,SRAStudy,BioProject,Study_Pubmed_id,ProjectID,Sample,BioSample,SampleType,TaxID,ScientificName,SampleName,g1k_pop_code,source,g1k_analysis_group,Subject_ID,Sex,Disease,Tumor,Affection_Status,Analyte_Type,Histological_Type,Body_Site,CenterName,Submission,dbgap_study_accession,Consent,RunHash,ReadHash\n",
      "SRR735691,2013-02-16,2013-02-16,14890,3305580,0,222,1,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR735691,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,EB5AD9CFD23FD4DDC17733913E95D003,42BEB11212FB634A454A1050BA36EDB9\n",
      "SRR747232,2013-02-19,2013-02-19,790903,270580865,0,342,145,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747232,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,73BCC76A8C83428156936A9E340AADB6,262713D63ED13145496C6579C67A9BD3\n",
      "SRR747256,2013-02-19,2013-02-19,524182,150277014,0,286,101,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747256,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,F22B1086B0A2D37EA5984E315D5199A7,4E6CB4272C39C64466DFC29E1725D4D2\n",
      "SRR747758,2013-02-20,2013-02-20,729194,241647757,0,331,133,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747758,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,2A3F5F819E841686C90E7A1EDB8F3C52,14269D403C302C1683E2BCFB8409A205\n",
      "SRR747759,2013-02-20,2013-02-20,475183,132503718,0,278,91,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747759,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,999F15957FD72D02642382F63E00536B,DE97F52EEB619FA76F391B97263EEF16\n",
      "SRR747760,2013-02-20,2013-02-20,1346039,466184361,0,346,286,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747760,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,66012EBF6A3A3BD0FDD7757F6A2F8B1F,814A169049FD1A9406EFC415E5E987FA\n",
      "SRR747761,2013-02-20,2013-02-20,1346301,385144557,0,286,227,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747761,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,CB80B91D67B5BD284CD872D30A45DCF8,B0EB49FC1C0C27F67CCB343026FCA557\n",
      "SRR747763,2013-02-20,2013-02-20,1269644,431233968,0,339,267,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747763,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,68BFB16E29987276732ADC45D74D0CC2,D09C794431A9E71209FA347A21EA08F2\n",
      "SRR747764,2013-02-20,2013-02-20,724207,123917539,0,171,86,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747764,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,14E85DF879510E0F981B4C9178DF46D4,3C68C75687CD661784084EAD6CF8256A\n",
      "SRR747766,2013-02-20,2013-02-20,1293020,400929357,0,310,263,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747766,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,C0540594E8A050F45182A0FDC9DD31E3,F0487A72D01CC104EFD90B06DE267F77\n",
      "SRR747767,2013-02-20,2013-02-20,1280726,443572549,0,346,278,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747767,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,6B61B24E6B75B49FE84A9A5D38B68FDE,13BB8B26A313577EEBD0790F62C755EE\n",
      "SRR747768,2013-02-26,2013-02-26,1186069,358665751,0,302,245,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747768,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,1531FD1DC44FC32252F2D55E018B6EFD,D9D9C1918150336B7D34BDD40A242B01\n",
      "SRR747785,2013-02-20,2013-02-20,1217068,396472345,0,325,257,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR747785,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,FFDDF3DB95ECDAC80B880ED167AE4DF1,923509ABE9E9DD17A3FA6D474787F4FB\n",
      "SRR765688,2013-02-26,2013-02-26,277826,71284046,0,256,45,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR765688,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,5BCE478D8F77B2B3E84498CD657B98F9,5ECB07090D470E01C135C770CA04F464\n",
      "SRR770500,2013-03-12,2013-03-12,999304,282759360,0,282,192,,https://sra-download.ncbi.nlm.nih.gov/srapub/SRR770500,SRX190717,,RNA-Seq,PCR,TRANSCRIPTOMIC,SINGLE,0,0,LS454,454 GS FLX Titanium,SRP015957,PRJNA176314,,176314,SRS366185,SAMN01737268,simple,9606,Homo sapiens,6,,,,,,,no,,,,,STANFORD UNIVERSITY,SRA058972,,public,4F8BA33596F993746A2E776D3E780388,2E15515BA1F287907BF3C9948D9F5447\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "esearch -db sra -query SRA058972 | efetch -format runinfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Munging the runinfo table\n",
    "\n",
    "- Once we have the runinfo table, the first column (run) can be extracted\n",
    "  - Python code for this is given in the exercise\n",
    "  - Some bash code is given below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "slideshow": {
     "slide_type": "subslide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR735691\n",
      "SRR747232\n",
      "SRR747256\n",
      "SRR747758\n",
      "SRR747759\n",
      "SRR747760\n",
      "SRR747761\n",
      "SRR747763\n",
      "SRR747764\n",
      "SRR747766\n",
      "SRR747767\n",
      "SRR747768\n",
      "SRR747785\n",
      "SRR765688\n",
      "SRR770500\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "esearch -db sra -query SRA058972 | efetch -format runinfo | cut -d ',' -f 1 | grep SRR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Downloading fastq files\n",
    "\n",
    "- We can pipe a list of accessions to `fastq-dump` in order to download the data\n",
    "\n",
    "```\n",
    "cat accessions.txt | xargs fastq-dump --outdir SRA058972\n",
    "```\n",
    "\n",
    "- If we had paired end reads (which we don't here), we would need two more options to `fastq-dump`:\n",
    "\n",
    "\n",
    "```\n",
    "cat accessions.txt | xargs fastq-dump --split-files --readids --outdir SRA058972\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
