{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IgBLAST for TCR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given the FASTA file `A2-i131.fasta`, use IgBLAST to assign germline V, D, and J segments, and post-process using Change-O."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtain reference sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Go to [http://www.imgt.org/vquest/refseqh.html](http://www.imgt.org/vquest/refseqh.html) and download TRBV, TRBD, and TRBJ sequences for humans in FASTA format. Save them as `IMGT_Human_TRBV.fasta` etc..\n",
    "\n",
    "(For advanced users, one could download the whole database from [here](http://www.imgt.org/download/GENE-DB/) and postprocess.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Converting IMGT FASTA files to IgBLAST databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 10/27/2016 07:30:54\n",
      "New DB name:   /home/simon/Projects/aairr16/solutions/database/human_trb_v\n",
      "New DB title:  database/human_trb_v\n",
      "Sequence type: Nucleotide\n",
      "Keep Linkouts: T\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 142 sequences in 0.027348 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 10/27/2016 07:30:54\n",
      "New DB name:   /home/simon/Projects/aairr16/solutions/database/human_trb_d\n",
      "New DB title:  database/human_trb_d\n",
      "Sequence type: Nucleotide\n",
      "Keep Linkouts: T\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 3 sequences in 0.00134206 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 10/27/2016 07:30:54\n",
      "New DB name:   /home/simon/Projects/aairr16/solutions/database/human_trb_j\n",
      "New DB title:  database/human_trb_j\n",
      "Sequence type: Nucleotide\n",
      "Keep Linkouts: T\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 16 sequences in 0.00300503 seconds.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# V-segment database\n",
    "perl ./edit_imgt_file.pl IMGT_Human_TRBV.fasta > database/human_trb_v\n",
    "makeblastdb -parse_seqids -dbtype nucl -in database/human_trb_v\n",
    "# D-segment database\n",
    "perl ./edit_imgt_file.pl IMGT_Human_TRBD.fasta > database/human_trb_d\n",
    "makeblastdb -parse_seqids -dbtype nucl -in database/human_trb_d\n",
    "# J-segment database\n",
    "perl ./edit_imgt_file.pl IMGT_Human_TRBJ.fasta > database/human_trb_j\n",
    "makeblastdb -parse_seqids -dbtype nucl -in database/human_trb_j"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use IgBLAST"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "IgBLAST has many options (see below) but the most important ones are as follows:\n",
    "\n",
    "- germline_db_V: the V gene database\n",
    "- germline_db_D: the D gene database\n",
    "- germline_db_J: the J gene database\n",
    "- auxiliary_data: contains annotations for the sequences\n",
    "- domain_system: the system used (e.g. imgt) for defining the domains\n",
    "- ig_seqtype: Ig or TCR\n",
    "- organism: e.g. human, mouse\n",
    "- outfmt: the output format; for postprocessing with ChangeO, has to be '7 std qseq sseq btop'\n",
    "- query: the input data in FASTA format\n",
    "- out: the output filename\n",
    "- num_threads: the number of threads to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USAGE\r\n",
      "  igblastn [-h] [-help] [-import_search_strategy filename]\r\n",
      "    [-export_search_strategy filename] [-germline_db_V germline_database_name]\r\n",
      "    [-num_alignments_V int_value] [-germline_db_V_seqidlist filename]\r\n",
      "    [-germline_db_D germline_database_name] [-num_alignments_D int_value]\r\n",
      "    [-germline_db_D_seqidlist filename]\r\n",
      "    [-germline_db_J germline_database_name] [-num_alignments_J int_value]\r\n",
      "    [-germline_db_J_seqidlist filename] [-auxiliary_data filename]\r\n",
      "    [-min_D_match min_D_match] [-D_penalty D_penalty]\r\n",
      "    [-organism germline_origin] [-domain_system domain_system]\r\n",
      "    [-ig_seqtype sequence_type] [-focus_on_V_segment] [-show_translation]\r\n",
      "    [-db database_name] [-dbsize num_letters] [-gilist filename]\r\n",
      "    [-seqidlist filename] [-negative_gilist filename]\r\n",
      "    [-entrez_query entrez_query] [-db_soft_mask filtering_algorithm]\r\n",
      "    [-db_hard_mask filtering_algorithm] [-subject subject_input_file]\r\n",
      "    [-subject_loc range] [-query input_file] [-out output_file]\r\n",
      "    [-evalue evalue] [-word_size int_value] [-gapopen open_penalty]\r\n",
      "    [-gapextend extend_penalty] [-perc_identity float_value]\r\n",
      "    [-qcov_hsp_perc float_value] [-xdrop_ungap float_value]\r\n",
      "    [-xdrop_gap float_value] [-xdrop_gap_final float_value]\r\n",
      "    [-searchsp int_value] [-max_hsps int_value] [-sum_statistics]\r\n",
      "    [-penalty penalty] [-reward reward] [-no_greedy] [-ungapped]\r\n",
      "    [-culling_limit int_value] [-best_hit_overhang float_value]\r\n",
      "    [-best_hit_score_edge float_value] [-window_size int_value]\r\n",
      "    [-off_diagonal_range int_value] [-lcase_masking] [-query_loc range]\r\n",
      "    [-strand strand] [-parse_deflines] [-outfmt format] [-show_gis]\r\n",
      "    [-num_descriptions int_value] [-num_alignments int_value]\r\n",
      "    [-line_length line_length] [-max_target_seqs num_sequences]\r\n",
      "    [-num_threads int_value] [-remote] [-version]\r\n",
      "\r\n",
      "DESCRIPTION\r\n",
      "   Nucleotide-Nucleotide BLAST for immunoglobulin sequences 2.2.29+\r\n",
      "\r\n",
      "OPTIONAL ARGUMENTS\r\n",
      " -h\r\n",
      "   Print USAGE and DESCRIPTION;  ignore all other parameters\r\n",
      " -help\r\n",
      "   Print USAGE, DESCRIPTION and ARGUMENTS; ignore all other parameters\r\n",
      " -version\r\n",
      "   Print version number;  ignore other arguments\r\n",
      "\r\n",
      " *** Input query options\r\n",
      " -query <File_In>\r\n",
      "   Input file name\r\n",
      "   Default = `-'\r\n",
      " -query_loc <String>\r\n",
      "   Location on the query sequence in 1-based offsets (Format: start-stop)\r\n",
      " -strand <String, `both', `minus', `plus'>\r\n",
      "   Query strand(s) to search against database/subject\r\n",
      "   Default = `both'\r\n",
      "\r\n",
      " *** General search options\r\n",
      " -db <String>\r\n",
      "   BLAST database name\r\n",
      "    * Incompatible with:  subject, subject_loc\r\n",
      " -out <File_Out>\r\n",
      "   Output file name\r\n",
      "   Default = `-'\r\n",
      " -evalue <Real>\r\n",
      "   Expectation value (E) threshold for saving hits \r\n",
      "   Default = `1e-15'\r\n",
      " -word_size <Integer, >=4>\r\n",
      "   Word size for wordfinder algorithm (length of best perfect match)\r\n",
      " -gapopen <Integer>\r\n",
      "   Cost to open a gap\r\n",
      " -gapextend <Integer>\r\n",
      "   Cost to extend a gap\r\n",
      " -penalty <Integer, <=0>\r\n",
      "   Penalty for a nucleotide mismatch\r\n",
      " -reward <Integer, >=0>\r\n",
      "   Reward for a nucleotide match\r\n",
      "\r\n",
      " *** BLAST-2-Sequences options\r\n",
      " -subject <File_In>\r\n",
      "   Subject sequence(s) to search\r\n",
      "    * Incompatible with:  db, gilist, seqidlist, negative_gilist,\r\n",
      "   db_soft_mask, db_hard_mask\r\n",
      " -subject_loc <String>\r\n",
      "   Location on the subject sequence in 1-based offsets (Format: start-stop)\r\n",
      "    * Incompatible with:  db, gilist, seqidlist, negative_gilist,\r\n",
      "   db_soft_mask, db_hard_mask, remote\r\n",
      "\r\n",
      " *** Formatting options\r\n",
      " -outfmt <String>\r\n",
      "   alignment view options:\r\n",
      "     3 = flat query-anchored, show identities,\r\n",
      "     4 = flat query-anchored, no identities,\r\n",
      "     7 = tabular with comment lines\r\n",
      "   \r\n",
      "   Options 7 can be additionally configured to produce\r\n",
      "   a custom format specified by space delimited format specifiers.\r\n",
      "   The supported format specifiers are:\r\n",
      "   \t    qseqid means Query Seq-id\r\n",
      "   \t       qgi means Query GI\r\n",
      "   \t      qacc means Query accesion\r\n",
      "   \t   qaccver means Query accesion.version\r\n",
      "   \t      qlen means Query sequence length\r\n",
      "   \t    sseqid means Subject Seq-id\r\n",
      "   \t sallseqid means All subject Seq-id(s), separated by a ';'\r\n",
      "   \t       sgi means Subject GI\r\n",
      "   \t    sallgi means All subject GIs\r\n",
      "   \t      sacc means Subject accession\r\n",
      "   \t   saccver means Subject accession.version\r\n",
      "   \t   sallacc means All subject accessions\r\n",
      "   \t      slen means Subject sequence length\r\n",
      "   \t    qstart means Start of alignment in query\r\n",
      "   \t      qend means End of alignment in query\r\n",
      "   \t    sstart means Start of alignment in subject\r\n",
      "   \t      send means End of alignment in subject\r\n",
      "   \t      qseq means Aligned part of query sequence\r\n",
      "   \t      sseq means Aligned part of subject sequence\r\n",
      "   \t    evalue means Expect value\r\n",
      "   \t  bitscore means Bit score\r\n",
      "   \t     score means Raw score\r\n",
      "   \t    length means Alignment length\r\n",
      "   \t    pident means Percentage of identical matches\r\n",
      "   \t    nident means Number of identical matches\r\n",
      "   \t  mismatch means Number of mismatches\r\n",
      "   \t  positive means Number of positive-scoring matches\r\n",
      "   \t   gapopen means Number of gap openings\r\n",
      "   \t      gaps means Total number of gaps\r\n",
      "   \t      ppos means Percentage of positive-scoring matches\r\n",
      "   \t    frames means Query and subject frames separated by a '/'\r\n",
      "   \t    qframe means Query frame\r\n",
      "   \t    sframe means Subject frame\r\n",
      "   \t      btop means Blast traceback operations (BTOP)\r\n",
      "   \t   staxids means unique Subject Taxonomy ID(s), separated by a ';'\r\n",
      "   \t\t\t (in numerical order)\r\n",
      "   \t sscinames means unique Subject Scientific Name(s), separated by a ';'\r\n",
      "   \t scomnames means unique Subject Common Name(s), separated by a ';'\r\n",
      "   \tsblastnames means unique Subject Blast Name(s), separated by a ';'\r\n",
      "   \t\t\t (in alphabetical order)\r\n",
      "   \tsskingdoms means unique Subject Super Kingdom(s), separated by a ';'\r\n",
      "   \t\t\t (in alphabetical order) \r\n",
      "   \t    stitle means Subject Title\r\n",
      "   \tsalltitles means All Subject Title(s), separated by a '<>'\r\n",
      "   \t   sstrand means Subject Strand\r\n",
      "   \t     qcovs means Query Coverage Per Subject\r\n",
      "   \t   qcovhsp means Query Coverage Per HSP\r\n",
      "   When not provided, the default value is:\r\n",
      "   'qseqid sseqid pident length mismatch gapopen gaps qstart qend sstart send\r\n",
      "   evalue bitscore', which is equivalent to the keyword 'std'\r\n",
      "   Default = `3'\r\n",
      " -show_gis\r\n",
      "   Show NCBI GIs in deflines?\r\n",
      " -num_descriptions <Integer, >=0>\r\n",
      "   Number of database sequences to show one-line descriptions for\r\n",
      "   Not applicable for outfmt > 4\r\n",
      "   Default = `10'\r\n",
      "    * Incompatible with:  max_target_seqs\r\n",
      " -num_alignments <Integer, >=0>\r\n",
      "   Number of database sequences to show alignments for\r\n",
      "   Default = `10'\r\n",
      "    * Incompatible with:  max_target_seqs\r\n",
      " -line_length <Integer, >=1>\r\n",
      "   Line length for formatting alignments\r\n",
      "   Not applicable for outfmt > 4\r\n",
      "   Default = `60'\r\n",
      "\r\n",
      " *** Query filtering options\r\n",
      " -lcase_masking\r\n",
      "   Use lower case filtering in query and subject sequence(s)?\r\n",
      "\r\n",
      " *** Restrict search or results\r\n",
      " -gilist <String>\r\n",
      "   Restrict search of database to list of GI's\r\n",
      "    * Incompatible with:  negative_gilist, seqidlist, remote, subject,\r\n",
      "   subject_loc\r\n",
      " -seqidlist <String>\r\n",
      "   Restrict search of database to list of SeqId's\r\n",
      "    * Incompatible with:  gilist, negative_gilist, remote, subject,\r\n",
      "   subject_loc\r\n",
      " -negative_gilist <String>\r\n",
      "   Restrict search of database to everything except the listed GIs\r\n",
      "    * Incompatible with:  gilist, seqidlist, remote, subject, subject_loc\r\n",
      " -entrez_query <String>\r\n",
      "   Restrict search with the given Entrez query\r\n",
      "    * Requires:  remote\r\n",
      " -db_soft_mask <String>\r\n",
      "   Filtering algorithm ID to apply to the BLAST database as soft masking\r\n",
      "    * Incompatible with:  db_hard_mask, subject, subject_loc\r\n",
      " -db_hard_mask <String>\r\n",
      "   Filtering algorithm ID to apply to the BLAST database as hard masking\r\n",
      "    * Incompatible with:  db_soft_mask, subject, subject_loc\r\n",
      " -perc_identity <Real, 0..100>\r\n",
      "   Percent identity\r\n",
      " -qcov_hsp_perc <Real, 0..100>\r\n",
      "   Percent query coverage per hsp\r\n",
      " -culling_limit <Integer, >=0>\r\n",
      "   If the query range of a hit is enveloped by that of at least this many\r\n",
      "   higher-scoring hits, delete the hit\r\n",
      "    * Incompatible with:  best_hit_overhang, best_hit_score_edge\r\n",
      " -best_hit_overhang <Real, (>0 and <0.5)>\r\n",
      "   Best Hit algorithm overhang value (recommended value: 0.1)\r\n",
      "    * Incompatible with:  culling_limit\r\n",
      " -best_hit_score_edge <Real, (>0 and <0.5)>\r\n",
      "   Best Hit algorithm score edge value (recommended value: 0.1)\r\n",
      "    * Incompatible with:  culling_limit\r\n",
      " -max_target_seqs <Integer, >=1>\r\n",
      "   Maximum number of aligned sequences to keep \r\n",
      "   Not applicable for outfmt <= 4\r\n",
      "   Default = `500'\r\n",
      "    * Incompatible with:  num_descriptions, num_alignments\r\n",
      "\r\n",
      " *** Statistical options\r\n",
      " -dbsize <Int8>\r\n",
      "   Effective length of the database \r\n",
      " -searchsp <Int8, >=0>\r\n",
      "   Effective length of the search space\r\n",
      " -max_hsps <Integer, >=0>\r\n",
      "   Set maximum number of HSPs per subject sequence to save (0 means no limit)\r\n",
      "   Default = `0'\r\n",
      " -sum_statistics\r\n",
      "   Use sum statistics\r\n",
      "\r\n",
      " *** Search strategy options\r\n",
      " -import_search_strategy <File_In>\r\n",
      "   Search strategy to use\r\n",
      "    * Incompatible with:  export_search_strategy\r\n",
      " -export_search_strategy <File_Out>\r\n",
      "   File name to record the search strategy used\r\n",
      "    * Incompatible with:  import_search_strategy\r\n",
      "\r\n",
      " *** Extension options\r\n",
      " -xdrop_ungap <Real>\r\n",
      "   X-dropoff value (in bits) for ungapped extensions\r\n",
      " -xdrop_gap <Real>\r\n",
      "   X-dropoff value (in bits) for preliminary gapped extensions\r\n",
      " -xdrop_gap_final <Real>\r\n",
      "   X-dropoff value (in bits) for final gapped alignment\r\n",
      " -no_greedy\r\n",
      "   Use non-greedy dynamic programming extension\r\n",
      " -ungapped\r\n",
      "   Perform ungapped alignment only?\r\n",
      " -window_size <Integer, >=0>\r\n",
      "   Multiple hits window size, use 0 to specify 1-hit algorithm\r\n",
      " -off_diagonal_range <Integer, >=0>\r\n",
      "   Number of off-diagonals to search for the 2nd hit, use 0 to turn off\r\n",
      "   Default = `0'\r\n",
      "\r\n",
      " *** Ig-BLAST options\r\n",
      " -germline_db_V <String>\r\n",
      "   Germline database name\r\n",
      " -num_alignments_V <Integer>\r\n",
      "   Number of Germline sequences to show alignments for\r\n",
      "   Default = `3'\r\n",
      " -germline_db_V_seqidlist <String>\r\n",
      "   Restrict search of germline database to list of SeqIds's\r\n",
      " -germline_db_D <String>\r\n",
      "   Germline database name\r\n",
      " -num_alignments_D <Integer>\r\n",
      "   Number of Germline sequences to show alignments for\r\n",
      "   Default = `3'\r\n",
      " -germline_db_D_seqidlist <String>\r\n",
      "   Restrict search of germline database to list of SeqIds's\r\n",
      " -germline_db_J <String>\r\n",
      "   Germline database name\r\n",
      " -num_alignments_J <Integer>\r\n",
      "   Number of Germline sequences to show alignments for\r\n",
      "   Default = `3'\r\n",
      " -germline_db_J_seqidlist <String>\r\n",
      "   Restrict search of germline database to list of SeqIds's\r\n",
      " -auxiliary_data <String>\r\n",
      "   File containing the coding frame start positions for sequences in germline\r\n",
      "   J database\r\n",
      " -min_D_match <Integer, >=5>\r\n",
      "   Required minimal number of D gene matches \r\n",
      " -D_penalty <Integer, (>-6 and <0)>\r\n",
      "   Penalty for a nucleotide mismatch in D gene\r\n",
      "   Default = `-4'\r\n",
      " -organism <String, `human', `mouse', `rabbit', `rat', `rhesus_monkey'>\r\n",
      "   The organism for your query sequence (i.e., human, mouse, etc.)\r\n",
      "   Default = `human'\r\n",
      " -domain_system <String, `imgt', `kabat'>\r\n",
      "   Domain system to be used for segment annotation\r\n",
      "   Default = `kabat'\r\n",
      " -ig_seqtype <String, `Ig', `TCR'>\r\n",
      "   Specify Ig or T cell receptor sequence\r\n",
      "   Default = `Ig'\r\n",
      " -focus_on_V_segment\r\n",
      "   Should the search only be for V segment (effective only for non-germline\r\n",
      "   database search using -db option)?\r\n",
      " -show_translation\r\n",
      "   Show translated alignments\r\n",
      "\r\n",
      " *** Miscellaneous options\r\n",
      " -parse_deflines\r\n",
      "   Should the query and subject defline(s) be parsed?\r\n",
      " -num_threads <Integer, >=1>\r\n",
      "   Number of threads (CPUs) to use in the BLAST search\r\n",
      "   Default = `1'\r\n",
      "    * Incompatible with:  remote\r\n",
      " -remote\r\n",
      "   Execute search remotely?\r\n",
      "    * Incompatible with:  gilist, seqidlist, negative_gilist, subject_loc,\r\n",
      "   num_threads\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!igblastn -help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Complete the following cell to run `A2-i131,fasta` against the TRB databases generated previously. Ensure that the outfmt term is '7 std qseq sseq btop', and save the output as `A2-i131.fmt7`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "igblastn \\\n",
    "    -germline_db_V database/human_trb_v \\\n",
    "    -germline_db_D database/human_trb_d \\\n",
    "    -germline_db_J database/human_trb_j \\\n",
    "    -auxiliary_data optional_file/human_gl.aux \\\n",
    "    -domain_system imgt -ig_seqtype TCR -organism human \\\n",
    "    -outfmt '7 std qseq sseq btop' \\\n",
    "    -query A2-i131.fasta \\\n",
    "    -out A2-i131.fmt7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Postprocess IgBLAST"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following cells postprocess the BLAST output using Change-O."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        START> MakeDB\n",
      "      ALIGNER> IgBlast\n",
      "ALIGN_RESULTS> SRR765688.fmt7\n",
      "     SEQ_FILE> SRR765688.fasta\n",
      "     NO_PARSE> False\n",
      " SCORE_FIELDS> True\n",
      "REGION_FIELDS> True\n",
      "\n",
      "PROGRESS> 07:54:44 [                    ]   0% (    0) 0.0 min\r",
      "PROGRESS> 07:54:44 [#                   ]   5% (  129) 0.0 min\r",
      "PROGRESS> 07:54:45 [##                  ]  10% (  258) 0.0 min\r",
      "PROGRESS> 07:54:46 [###                 ]  15% (  387) 0.0 min\r",
      "PROGRESS> 07:54:46 [####                ]  20% (  516) 0.0 min\r",
      "PROGRESS> 07:54:47 [#####               ]  25% (  645) 0.1 min\r",
      "PROGRESS> 07:54:47 [######              ]  30% (  774) 0.1 min\r",
      "PROGRESS> 07:54:48 [#######             ]  35% (  903) 0.1 min\r",
      "PROGRESS> 07:54:49 [########            ]  40% (1,032) 0.1 min\r",
      "PROGRESS> 07:54:50 [#########           ]  45% (1,161) 0.1 min\r",
      "PROGRESS> 07:54:50 [##########          ]  50% (1,290) 0.1 min\r",
      "PROGRESS> 07:54:51 [###########         ]  55% (1,419) 0.1 min\r",
      "PROGRESS> 07:54:52 [############        ]  60% (1,548) 0.1 min\r",
      "PROGRESS> 07:54:52 [#############       ]  65% (1,677) 0.1 min\r",
      "PROGRESS> 07:54:53 [##############      ]  71% (1,806) 0.2 min\r",
      "PROGRESS> 07:54:53 [###############     ]  76% (1,935) 0.2 min\r",
      "PROGRESS> 07:54:54 [################    ]  81% (2,064) 0.2 min\r",
      "PROGRESS> 07:54:55 [#################   ]  86% (2,193) 0.2 min\r",
      "PROGRESS> 07:54:55 [##################  ]  91% (2,322) 0.2 min\r",
      "PROGRESS> 07:54:56 [################### ]  96% (2,451) 0.2 min\r",
      "PROGRESS> 07:54:56 [####################] 100% (2,561) 0.2 min\n",
      "\n",
      "OUTPUT> SRR765688_db-pass.tab\n",
      "  PASS> 2553\n",
      "  FAIL> 8\n",
      "   END> MakeDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "MakeDb.py igblast -i A2-i131.fmt7 -s A2-i131.fasta -r IMGT_Human_TRB[VDJ].fasta \\\n",
    "    --regions --scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    START> ParseDb\n",
      "  COMMAND> split\n",
      "     FILE> SRR765688_db-pass.tab\n",
      "    FIELD> FUNCTIONAL\n",
      "NUM_SPLIT> None\n",
      "\n",
      "PROGRESS> 07:55:05 [                    ]   0% (    0) 0.0 min\r",
      "PROGRESS> 07:55:05 [#                   ]   5% (  128) 0.0 min\r",
      "PROGRESS> 07:55:05 [##                  ]  10% (  256) 0.0 min\r",
      "PROGRESS> 07:55:05 [###                 ]  15% (  384) 0.0 min\r",
      "PROGRESS> 07:55:05 [####                ]  20% (  512) 0.0 min\r",
      "PROGRESS> 07:55:05 [#####               ]  25% (  640) 0.0 min\r",
      "PROGRESS> 07:55:05 [######              ]  30% (  768) 0.0 min\r",
      "PROGRESS> 07:55:05 [#######             ]  35% (  896) 0.0 min\r",
      "PROGRESS> 07:55:05 [########            ]  40% (1,024) 0.0 min\r",
      "PROGRESS> 07:55:05 [#########           ]  45% (1,152) 0.0 min\r",
      "PROGRESS> 07:55:05 [##########          ]  50% (1,280) 0.0 min\r",
      "PROGRESS> 07:55:05 [###########         ]  55% (1,408) 0.0 min\r",
      "PROGRESS> 07:55:05 [############        ]  60% (1,536) 0.0 min\r",
      "PROGRESS> 07:55:05 [#############       ]  65% (1,664) 0.0 min\r",
      "PROGRESS> 07:55:05 [##############      ]  70% (1,792) 0.0 min\r",
      "PROGRESS> 07:55:05 [###############     ]  75% (1,920) 0.0 min\r",
      "PROGRESS> 07:55:05 [################    ]  80% (2,048) 0.0 min\r",
      "PROGRESS> 07:55:05 [#################   ]  85% (2,176) 0.0 min\r",
      "PROGRESS> 07:55:05 [##################  ]  90% (2,304) 0.0 min\r",
      "PROGRESS> 07:55:05 [################### ]  95% (2,432) 0.0 min\r",
      "PROGRESS> 07:55:05 [####################] 100% (2,553) 0.0 min\n",
      "\n",
      "OUTPUT1> SRR765688_db-pass_FUNCTIONAL-F.tab\n",
      "OUTPUT2> SRR765688_db-pass_FUNCTIONAL-T.tab\n",
      "RECORDS> 2553\n",
      "  PARTS> 2\n",
      "    END> ParseDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "ParseDb.py split -d A2-i131_db-pass.tab -f FUNCTIONAL"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
